pybamm/expression_tree/symbol.py

```diff
@@ -417,14 +419,12 @@ def scale(self):
     def reference(self):
         return self._reference
 
+    @cache
     def __eq__(self, other):
-        try:
-            return self._id == other._id
-        except AttributeError:
-            if isinstance(other, numbers.Number):
-                return self._id == pybamm.Scalar(other)._id
-            else:
-                return False
+        if isinstance(other, numbers.Number):
+            return self._id == pybamm.Scalar(other)._id
+        else:
+            return False
```

Comment:
this one should not be edited, the `AttributeError` is doing something different here (you can tell because there is no assignment happening)